### ISSUE TITLE
Should we honour @WebServlet?

### ISSUE THREAD

**Issue by wilkinsona** [spring-projects/spring-boot#2290]
`ServletContextInitializerBeans` finds all beans that are a `Servlet` and creates a `ServletRegistrationBean` for each of them. We could perhaps, [as this user expected](http://stackoverflow.com/questions/27785935/spring-boot-does-not-honor-webservlet), look for `@WebServlet` on such a bean to determine, among other things, the url patterns to which the servlet should be mapped.

**Comment by philwebb** [comment]
There is also `@WebFilter` to consider. My initial feeling is that it might be confusing to try and support these annotations directly. We already made the decision not to support `@HandlesTypes` and `WebApplicationInitializer` and this one feels like it's in a similar category. Perhaps we could add our own equivalent annotations that are also `@Component` annotated?

**Comment by wilkinsona** [comment]
I think both `@WebServlet` and `@WebFilter` are slightly different. We support both Servlets and Filters so this is about how the user configures them. The ideal, imo, would be able to use `@WebServlet` and `@WebFilter` and _not_ have to also use `@Component`. Failing that, our own `@WebServlet` and `@WebFilter` annotations that are @Component-annotated would still be worthwhile, imo.

**Comment by miki4you** [comment]
How can one use a third party library that is based on those annoations with Spring Boot without having to search the source code of the library and add each Filter/Servlet manually ?  Is there a way to tell the web app container to search for these annotations ?

**Comment by wilkinsona** [comment]
@miki4you The embedded containers don't perform any scanning for `@WebServlet` or `@WebFilter`. Can you point us to some libraries that make use of them and that are causing you a problem?

**Comment by joensson** [comment]
@miki4you Though not Spring Boot specific it is possible to configure the embedded Tomcat do the scanning for @WebServlet and @WebFilter. 

See the accepted solution on SO for how to enable it - the accepted solution contains the solution for both embedded Tomcat 7 and embedded Tomcat 8 http://stackoverflow.com/questions/11669507/embedded-tomcat-7-servlet-3-0-annotations-not-working 

Based on that answer I think you can achieve the same in Spring Boot by using the TomcatEmbeddedServletContainerFactory - something along these lines (not tested - pseudo code :)):

    @Bean
    public TomcatEmbeddedServletContainerFactory tomcatFactory() {
        return new TomcatEmbeddedServletContainerFactory() {
            @Override
            protected void postProcessContext(Context context) {
                WebResourceSet myCustomResourceSet = new JarResourceSet();
//or DirResourceSet, or FileResourceSet or JarWarResourceSet 
                context.getResources().addPreResources(myCustomResourceSet);
            }
        };
    }

**Comment by miki4you** [comment]
@wilkinsona  i wanted to try out https://github.com/stagemonitor/stagemonitor with spring boot , followed the configuration tutorial and when the app started nothing happend. i've taken a look at the source and saw the heavy use of WebFilter/WebServlet. 

Then i started searching the web for a solution , tought it was a spring boot problem. But as @joensson  pointed out it seems to be related with the embedded web container.  Thank you @joensson for the quick answer. i'm going to try it out as soon as possible.

**Comment by miki4you** [comment]
I 've just tried to get it working with the help form your example. To first thing i noticed is that the call to
```
 context.getResources()
``` 
returns null, which means the WebResource Set can not be added. 

In the Stackoverflow example , they are working with a "StandardContext" from de deployed wepapp. 
I searched the Embedded Tomcat Documentation to find an example to create a JarStandartContext  to code looks like this 

```
    @Bean
    public TomcatEmbeddedServletContainerFactory tomcatFactory() {
        return new TomcatEmbeddedServletContainerFactory() {

            @Override
            protected void postProcessContext(Context context) {

                context.setAddWebinfClassesResources(true);

                WebResourceRoot resources = new StandardRoot();
                WebResourceSet myCustomResourceSet = new JarResourceSet(resources,"/","/","/");
                resources.addPreResources(myCustomResourceSet);
                context.setResources(resources);

            }
        };
    }
``` 

But now the app wont start up with the Exception 
org.springframework.context.ApplicationContextException: Unable to start embedded container;

**Comment by 8a8aleo** [comment]
As a long time Spring user I am disappointed that I needed to add spring specific code just to get a servlet up and running. Not a good sign for things to come imho.

**Comment by dsyer** [comment]
Thank you for your loyalty. How about proposing something concrete, like maybe in a pull request?

**Comment by miki4you** [comment]
@8a8aleo have you got the automatic scanning of servelt annotations working ? as you said i had to declare every single servelt/weblistener as a bean

**Comment by saszjt** [comment]
Running into the same thing with a dependent jar which contains @WebFilter annotations which are not getting picked up by Spring Boot.  Was quite difficult to track down why things were not working and while I can work around by registering filters for this jar myself it leaves the door open for breaking things any time I update dependencies if someone adds a new @WebFilter or @WebServlet annotation.

**Comment by felixbarny** [comment]
@miki4you did you get the annotation scanning to work?

Is that something you spring-boot guys would accept as a pull request?

Does someone have I idea for a more robust automatic filter registration that I could use in https://github.com/stagemonitor/stagemonitor?

**Comment by miki4you** [comment]
no i did not ... the submitted pseudo code above did not work after my modifications .. maybe somone with more skills on the embedded container could solve this problem.

for stagemonitor i decalred every servlet as a spring bean.. got it to work for demo purposes but then stopped playing with it for the moment because the solution with declarding every single servlet just seems odd. 

as far as i know , spring boot does not have to add any additional code if we could turn on the annoations scanning for specific packages

**Comment by thomasdarimont** [comment]
Until spring-boot provides support for automatically picking up ```@WebServlet```, ```@WebListener```and ```@WebFilter```you could use component scanning to discover those components yourself and dynamically create appropriate BeanDefinitions.

Here is an example for that: https://github.com/thomasdarimont/spring-boot-jee-web-example

**Comment by felixbarny** [comment]
But what if someone uses Spring Boot to create a war that is deployed to a 'classical' tomcat installation? Will the Servlets be picked up twice?

**Comment by thomasdarimont** [comment]
Yes - but in this case you don't need the workaround in the first place, right? :)

**Comment by wilkinsona** [comment]
@thomasdarimont Nice example, thanks!

@felixbarny If you're creating an executable war, i.e. one that will both be run with `java -jar` and deployed to a servlet container, you could limit the functionality to the embedded case by checking that the application context is an `EmbeddedWebApplicationContext` and that `getEmbeddedServletContainer()` does not return `null`.

**Comment by yacota** [comment]
+1 to add @thomasdarimont  {{JeeComponentsBeanFactoryPostProcessor }} to spring-boot 
Thanks

**Comment by felixbarny** [comment]
I guess the long-term solution would rather be to configure the embedded tomcat to pick up those `@Web...`-annotated classes. But I didn't have luck to find the right switch.. Any tomcat gurus around here?

**Comment by wilkinsona** [comment]
@felixbarny On first impressions at least, I prefer using Spring's component scanning machinery (as suggested by Thomas) as it'll work across embedded containers, i.e. we hopefully won't need different implementations for Jetty, Tomcat, and Undertow.

**Comment by felixbarny** [comment]
Good point, but in his example, you need to specify the package(s) to scan. What would the default be? To scan all packages like a non-embedded servlet container does?

**Comment by wilkinsona** [comment]
Given that this problem doesn't appear to affect many people, and that the performance cost of scanning every package may be quite high, I imagine that we'll make it opt-in and allow the packages to scan to be configured when you do so.

**Comment by felixbarny** [comment]
Another solution for stagemonitor would be to use a `javax.servlet.ServletContainerInitializer` instead of the `@Web*` annotations. Unfortunately, boot doesn't seem to support these (see #321). Is there a chance you could rethink that? It seems to me like boot is not compliant with the servlet spec at this point...

**Comment by felixbarny** [comment]
friendly ping :wink:

**Comment by dsyer** [comment]
Friendly reminder that this issue is still open and marked for 1.3.0.RC1. If you want to send a pull request that will help get it prioritised.

**Comment by felixbarny** [comment]
I'm willing to do that, but before I can code something I need more information (see also my questions in #321) and we probably have to agree on something.

**Comment by dsyer** [comment]
#321 has to stay orthogonal (i.e. whatever we introduce for `@WebServlet` processing has to not be a `ServletContainerInitializer`). That's the only thing I think we absolutely have to agree on. The rest will just be implementation detail (like whether it is on or off by default, and whether to require the user to list the packages to scan - my personal favourite answer to those questions is "off" and "yes", but I think we could adapt whatever you do to accommodate if the consensus was different).

**Comment by philwebb** [comment]
#3216 also discusses `@HandlesTypes`. I think we might need to defer this to 1.4 for now.

**Comment by dsyer** [comment]
On balance I would rather do something on this one now and defer the others.  I think whatever support, if any, we provide for SCI (see #321) will have to be hedged and heavily surrounded by optionality. Maybe it would it be possible to add support for `@WebServlet` if it was a `@Bean` (putting the onus on users to opt in but not making it over difficult if they know what they want to achieve)?

**Comment by philwebb** [comment]
Moving it back

**Comment by wilkinsona** [comment]
An implementation of this is available [here](https://github.com/wilkinsona/spring-boot/tree/gh-2290). It adds a new annotation, `@ServletComponentScan`, that can be used to switch on scanning for `@WebFilter`, `@WebListener`, and `@WebServlet` annotated classes.

**Comment by wilkinsona** [comment]
Closed by 003268fb4e
